doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,1399 @@
+
+More details about this import and its behavior can be found in [PEP 563](https://peps.python.org/pep-0563/)/[649](https://peps.python.org/pep-0649/).
+
+
+### Use type aliases to create readable names for types
+
+Use a type alias to create more readable types or convey intent.
+
+```python
+from typing import List, Tuple, Dict
+
+Coordinate = Tuple[float, float]
+Route = List[Coordinate]
+TrailName = str
+Trails = Dict[TrailName, Route]
+
+
+def get_trails(state) -> Trails:
+    ...
+
+# vs.
+
+def get_trails(state) -> Dict[str, List[Tuple[float, float]]]:
+    ...
+```
+
+If it's possible that a type alias could be confused with a global assignment, e.g. `x = 1`, you can use `typing.TypeAlias`.
+This removes ambiguity for the type checker and indicates this isn't a normal variable assignment.
+
+```python
+# from typing import TypeAlias Python >=3.10
+from typing_extensions import TypeAlias
+
+x: TypeAlias = 1
+```
+
+> Note that TypeAlias is backported to older versions of Python by using typing_extensions.
+
+### Use typing.overload to overload a function
+
+`typing.overload` allows for annotating different combinations of typed arguments for a function. This can be useful when the
+return type might depend on the type or types of parameters.
+
+Consider a function which takes in different input types which then inform the output type:
+
+```python
+from typing import Union
+
+
+def analyze_text(text: str, analysis_kind: Union[SentimentAnalysis, EntityRecognition, LanguageDetection]) -> Union[
+    SentimentResult, EntityRecognitionResult, LanguageDetectionResult]:
+    return _analyze(text, analysis_kind)
+```
+
+We can annotate this function with overloads to help inform the type checker which input maps to
+which return type.
+
+```python
+from typing import overload, Union
+
+
+@overload
+def analyze_text(text: str, analysis_kind: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: SentimentAnalysis) -> SentimentResult:
+    ...
+
+# actual implementation
+def analyze_text(text: str, analysis_kind: Union[SentimentAnalysis, EntityRecognition, LanguageDetection]) -> Union[
+    SentimentResult, EntityRecognitionResult, LanguageDetectionResult]:
+    return _analyze(text, analysis_kind)
+```
+
+Running mypy shows that the correct return type is inferred from the input type:
+
+```python
+result = analyze_text("my text", LanguageDetection())
+reveal_type(result)
+```
+
+`main.py:28: note: Revealed type is "__main__.LanguageDetectionResult"`
+
+`typing.overload` can also be used when there are a different number of typed arguments. We can take the above example and
+imagine that `analyze_text` has an additional overload which does not take an `analysis_kind` (note that we switched to accepting `*args` and `**kwargs` in the actual implementation for this case):
```

Comment:
Maybe we could explicitly mention here that overloads can be used to present optional parameters as required, or vice versa, in the particular overload scenario.